orca/service_impl.go

```diff
@@ -0,0 +1,173 @@
+func Register(s grpc.ServiceRegistrar, opts ServerOptions) (*Server, error) {
+	// TODO(easwars): Once the generated pb.gos are updated, we wont need this
+	// type assertion any more, since the new versions accept a
+	// grpc.ServiceRegistrar in their registration functions.
+	srv, ok := s.(*grpc.Server)
+	if !ok {
+		return nil, fmt.Errorf("concrete type of provided grpc.ServiceRegistrar is %T, only supported type is %T", s, &grpc.Server{})
+	}
+	server := NewServer(opts)
+	v3orcaservicegrpc.RegisterOpenRcaServiceServer(srv, server)
+	return server, nil
+}
+
+// NewServer returns a new ORCA service implementation configured using the
+// provided options. This can be used to customize the registration of the
+// returned ORCA server. Most usages should prefer to use Register instead.
+func NewServer(opts ServerOptions) *Server {
+	if opts.MinReportingInterval < internal.MinORCAReportingInterval {
+		opts.MinReportingInterval = internal.MinORCAReportingInterval
+	}
+	return &Server{
+		minReportingInterval: opts.MinReportingInterval,
+		recorder:             newMetricRecorder(),
+	}
+}
+
+func (s *Server) determineReportingInterval(req *v3orcaservicegrpc.OrcaLoadReportRequest) time.Duration {
+	if req.GetReportInterval() == nil {
+		return s.minReportingInterval
+	}
+	interval := req.GetReportInterval()
+	if err := interval.CheckValid(); err != nil {
+		logger.Warningf("Received reporting interval %q is invalid: %v. Using default: %s", interval, s.minReportingInterval)
+		return s.minReportingInterval
+	}
+	return interval.AsDuration()
+}
+
+func (s *Server) sendMetricsResponse(stream v3orcaservicegrpc.OpenRcaService_StreamCoreMetricsServer) error {
+	// The oobRecorder global singleton is guaranteed to be initialized at this
+	// point because registration of this service implementation and the
+	// initialization of the former take place in the same function,
+	// EnableOutOfBandMetricsReporting().
+	resp := s.recorder.toLoadReportProto()
+	return stream.Send(resp)
+}
+
+// StreamCoreMetrics streams custom backend metrics injected by the server
+// application.
+func (s *Server) StreamCoreMetrics(req *v3orcaservicepb.OrcaLoadReportRequest, stream v3orcaservicepb.OpenRcaService_StreamCoreMetricsServer) error {
+	// Send one response immediately before creating and waiting for the timer
+	// to fire.
+	if err := s.sendMetricsResponse(stream); err != nil {
+		return err
+	}
+
+	ticker := time.NewTicker(s.determineReportingInterval(req))
+	defer ticker.Stop()
+
+	for {
+		// Send a response containing the currently recorded metrics
+		select {
+		case <-stream.Context().Done():
+			return status.Error(codes.Canceled, "Stream has ended.")
+		case <-ticker.C:
+			if err := s.sendMetricsResponse(stream); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// SetUtilizationMetric records a measurement for a utilization metric uniquely
+// identifiable by name.
+func (s *Server) SetUtilizationMetric(name string, val float64) {
```

Comment:
Random idea (didn't think it through; could be bad): make `s.recorder` implement these methods and embed.

Reply:
Done.

Didn't do the same for the call metrics recorder type since it does not use all methods of the underlying recorder.

Given that `Service` is exported, does embedding the `metricRecorder` type present any other issues in terms of people being able to use reflection and do some weird stuff with the internals? Or do we just say, if you want to shoot yourself in the foot, go ahead and do it?